### DISCUSSION TITLE
Using HTTPS

### DISCUSSION THREAD

**Discussion by CenTdemeern1** [rust3ds/ctru-rs#199] [Q&A]
I'd like to perform an HTTPS request to an API somewhere from my code.
I've already tried using HTTP over TCP manually to avoid diving into SSL/TLS and external crates/libraries, but the API unfortunately requires HTTPS.

I've considered & tried the following:
- I've tried reqwest (doesn't work because of the lack of openssl)
- I've tried curl-rust (can't seem to compile, but this could maybe work if it were possible to make it use the [3ds-curl dkp portlib](https://github.com/devkitPro/pacman-packages/tree/master/3ds/curl)?)
- I've considered using bindgen with the 3ds-curl portlib, but I don't think I'm competent enough to make that link & build correctly

(Note that I've already tried the different backend crate features for the crates I've mentioned)

Does anyone have any ideas/suggestions? Help would be appreciated

**Comment by FenrirWolf** [comment]
I'm doing some investigation into getting `curl-rust` to link to the 3ds curl portlib. So far the two blockers I've encountered are 1) `curl-rust` always tries to link to OpenSSL on unix-like platforms (which the 3ds is considered as due to it having a number of unix-like compatibility shims) and 2) `curl-rust` depends on the `socket2` crate, which currently fails to compile for the 3ds due to various `libc` errors. The first problem is easy enough to solve with the right conditional compilation flags, but the second one is more involved. I'll see if I can fix everything locally and then submit PRs to the upstream repositories if I'm successful

**Reply by FenrirWolf** [reply]
Okay, I've got forks for both [curl-rust](https://github.com/FenrirWolf/curl-rust/tree/horizon) and [socket2](https://github.com/FenrirWolf/socket2/tree/horizon). If you want to try it out, you need a `Cargo.toml` that looks something like [this](https://gist.github.com/FenrirWolf/d49bba781a54d104c838d6bddc61b036) and you'll also a `.cargo/config.toml` that looks like [this](https://gist.github.com/FenrirWolf/ef518c7349cd80ac86e9319b8974c1a5) in order to link devkitPro's version of `libcurl`.

But although this lets us get past the compile and linking stages, I'm not sure if everything is actually working properly yet. When I run [this test program](https://gist.github.com/FenrirWolf/843b9caf4d1e5417add13fb8617b9135), I get the following panic message:

![2024-12-04_01-33-44 819_bot](https://github.com/user-attachments/assets/e289611c-ef51-49f6-84b6-275500db59ab)

So I guess there's still something that needs to be fixed somewhere.

**Reply by CenTdemeern1** [reply]
Oooh, great work! I'll have a look at this in a bit to see if I can help

**Reply by CenTdemeern1** [reply]
I had a look and this is all going over my head so I'm not sure what to do
What I *do* know is that zlib doesn't want to compile because of a familiar error I've seen before somewhere
https://gist.github.com/CenTdemeern1/98e4029247d57ba818b647a484bbedef
Any guesses as to what could be causing this?

**Reply by FenrirWolf** [reply]
Do you have the `3ds-zlib` package installed from `dkp-pacman`? I'm not sure if that will solve your issue but that's the first thing I would check for.

**Reply by FenrirWolf** [reply]
Either way, I still haven't figured out why the `curl-rust` example gives me that error. I compiled and ran an equivalent C program on my 3DS and it runs just fine, so as far as I can tell the problem only seems to happen when calling `3ds-curl` from Rust specifically. But now the question is why?

Some part of me is curious if https://github.com/rust3ds/ctru-rs/issues/174 might be the culprit, but I'm not familiar enough with socket programming to really know that for sure. @Meziu @adryzz Do you think that might have something to do with things or am I barking up the wrong tree here?

**Reply by CenTdemeern1** [reply]
> Do you have the `3ds-zlib` package installed from `dkp-pacman`? I'm not sure if that will solve your issue but that's the first thing I would check for.

I'll double check in a bit, but I remember checking my portlibs folder specifically for this reason and zlib being present.

**Reply by adryzz** [reply]
not sure right now, but I'll check it out. my 3DS' SD card died and i still haven't replaced it, so i haven't been active here for that reason, but I'll look into it later today.

**Reply by CenTdemeern1** [reply]
> > Do you have the `3ds-zlib` package installed from `dkp-pacman`? I'm not sure if that will solve your issue but that's the first thing I would check for.
> 
> I'll double check in a bit, but I remember checking my portlibs folder specifically for this reason and zlib being present.

Yep, I do have it
```zsh
% dkp-pacman -Q | grep zlib
3ds-zlib 1.3-1
```

**Reply by FenrirWolf** [reply]
And did you set up the [.cargo/config.toml ](https://gist.github.com/FenrirWolf/ef518c7349cd80ac86e9319b8974c1a5) too? That file overrides the build script for `curl-rust` and should take care of linking `libcurl` and `libz` for you.

**Reply by CenTdemeern1** [reply]
> And did you set up the [.cargo/config.toml ](https://gist.github.com/FenrirWolf/ef518c7349cd80ac86e9319b8974c1a5) too? That file overrides the build script for `curl-rust` and should take care of linking `libcurl` and `libz` for you.

[Yes, I did](https://github.com/CenTdemeern1/misskey-3ds/commit/247ce0f1e0fc5faa8f8664f5e22526e8936d595d#diff-9a4f3e4537ebb7474452d131b0d969d89a51286f4269aac5ef268e712be17268)

**Comment by adryzz** [comment]
reqwest should work with rustls... not sure what's the blocker there since the sockaddr_in issue is resolved now, i'll check back when i got some time, but yeah

**Reply by CenTdemeern1** [reply]
I remember trying rustls and it not being able to compile but I don't remember looking into if I can get it to work with the mbedtls portlib

**Reply by adryzz** [reply]
alright, the issue isn't in `rustls`, but in `mio`, i have a year old patch to make `mio` work on the 3ds, but i have to update it, and it'll take a while before it gets merged upstream, as the 3DS is a complex target
